File: packages/i18nify-js/src/modules/phoneNumber/formatPhoneNumber.ts
import { withErrorBoundary } from '../../common/errorBoundary';
import { CountryCodeType } from '../types/geo';
import { PHONE_FORMATTER_MAPPER } from './data/phoneFormatterMapper';
import { detectCountryAndDialCodeFromPhone, cleanPhoneNumber } from './utils';

// Formats a provided phone number according to the predefined format for a specific country code, or auto-detects the country code and formats the number accordingly.
const formatPhoneNumber = (
  phoneNumber: string | number,
  countryCode?: CountryCodeType,
): string => {
  // Throw errors if phoneNumber is invalid
  if (!phoneNumber) throw new Error('Parameter `phoneNumber` is invalid!');

  // Convert phoneNumber to string and clean it by removing non-numeric characters
  phoneNumber = phoneNumber.toString();
  phoneNumber = cleanPhoneNumber(phoneNumber);

  // Detect or validate the country code
  countryCode = (
    countryCode && countryCode in PHONE_FORMATTER_MAPPER
      ? countryCode
      : detectCountryAndDialCodeFromPhone(phoneNumber).countryCode
  ) as CountryCodeType;

  // Fetch the pattern for the countryCode from the PHONE_FORMATTER_MAPPER
  const pattern = PHONE_FORMATTER_MAPPER[countryCode];

  if (!pattern) return phoneNumber;

  // Count the number of 'x' characters in the format pattern
  let charCountInFormatterPattern = 0;
  for (let i = 0; i < pattern.length; i++) {
    if (pattern[i] === 'x') {
      charCountInFormatterPattern++;
    }
  }

  // Calculate the difference between phoneNumber length and 'x' characters count in pattern
  const diff = phoneNumber.length - charCountInFormatterPattern;
  // Extract the phoneNumber without the prefix
  const phoneNumberWithoutPrefix = phoneNumber.slice(diff);
  const formattedNumber: string[] = [];
  let numberIndex = 0;

  // Loop through the pattern to format the phoneNumber
  for (let i = 0; i < pattern.length; i++) {
    const patternChar = pattern[i];
    if (patternChar === 'x') {
      // Insert phoneNumber digits at 'x' positions
      if (numberIndex < phoneNumberWithoutPrefix.length) {
        formattedNumber.push(phoneNumberWithoutPrefix[numberIndex]);
        numberIndex++;
      }
    } else {
      // Insert non-digit characters from the pattern
      formattedNumber.push(patternChar);
    }
  }

  // Join the formattedNumber array to create the formattedPhoneNumber without prefix
  const formattedPhoneNumberWithoutPrefix = formattedNumber.join('');
  // Combine the prefix and formattedPhoneNumberWithoutPrefix
  const formattedPhoneNumberWithPrefix =
    phoneNumber.slice(0, diff) + ' ' + formattedPhoneNumberWithoutPrefix;

  // Return the formattedPhoneNumber with prefix after trimming whitespace
  return formattedPhoneNumberWithPrefix.trim();
};

export default withErrorBoundary<typeof formatPhoneNumber>(formatPhoneNumber);

File: packages/i18nify-go/modules/phonenumber/format_phone_number.go
package phonenumber

import (
	"fmt"
	"strings"
)

// FormatPhoneNumber formats phoneNumber according to the format template stored
// for countryCode in the dataset. When countryCode is empty or absent from the
// dataset, the country is auto-detected from the number's dial code prefix.
//
// Algorithm (mirrors the JS formatPhoneNumber function exactly):
//  1. Validate and clean the input.
//  2. Resolve the effective country code.
//  3. Count the 'x' placeholders in the format template.
//  4. Slice off everything after the diff = len(phone) - xCount characters
//     as the "prefix" (e.g. "+91"), and the remainder as the digits to format.
//  5. Walk the template: replace each 'x' with the next digit, copy separators
//     literally, then join prefix + " " + formatted and trim whitespace.
func FormatPhoneNumber(phoneNumber string, countryCode string) (string, error) {
	if phoneNumber == "" {
		return "", fmt.Errorf(
			"parameter 'phoneNumber' is invalid! The received value was: %s. Please ensure you provide a valid phone number",
			phoneNumber,
		)
	}

	cleaned := cleanPhoneNumber(phoneNumber)

	// Resolve effective country code.
	activeCC := countryCode
	if GetCountryTeleInformation(activeCC).Format == "" {
		activeCC, _ = detectCountryAndDialCodeFromPhone(cleaned)
	}

	info := GetCountryTeleInformation(activeCC)
	pattern := info.Format

	// No format pattern available — return the cleaned number as-is.
	if pattern == "" {
		return cleaned, nil
	}

	// Count 'x' placeholders in the pattern.
	xCount := strings.Count(pattern, "x")

	// diff is the length of the prefix that sits before the formatted segment.
	// e.g. for "+917394926646" (len=13) and pattern "xxxx xxxxxx" (10 x's): diff=3.
	diff := len([]rune(cleaned)) - xCount
	if diff < 0 {
		diff = 0
	}
	runes := []rune(cleaned)
	phoneWithoutPrefix := runes[diff:] // digits that map to 'x' slots

	// Fill the pattern.
	result := make([]rune, 0, len([]rune(pattern)))
	digitIdx := 0
	for _, ch := range pattern {
		if ch == 'x' {
			if digitIdx < len(phoneWithoutPrefix) {
				result = append(result, phoneWithoutPrefix[digitIdx])
				digitIdx++
			}
		} else {
			result = append(result, ch)
		}
	}

	// Prefix (e.g. "+91") + space + formatted segment, trimmed.
	prefix := string(runes[:diff])
	formatted := strings.TrimSpace(prefix + " " + string(result))
	return formatted, nil
}
